src/issue.rs

```diff
@@ -31,82 +30,525 @@ fn issue_actor_and_project(
 				.and_then(|issue_event| {
 					if issue_event.event == Some(github::Event::AddedToProject)
 					{
-						issue_event.project_card.as_ref().and_then(|card| {
-							card.project_url.as_ref().map(|project_url| {
-								(issue_event.actor.clone(), project_url)
-							})
+						issue_event.project_card.as_ref().map(|card| {
+							(issue_event.actor.clone(), card.clone())
 						})
 					} else {
 						None
 					}
 				})
-				.and_then(|(actor, project_url)| {
-					github_bot
-						.get(project_url)
-						.ok()
-						.map(|project| (actor, project))
-				})
 		})
 }
 
-pub fn handle_issue(
-	_db: &DB,
+fn author_admin_attach_only_project(
+	db_entry: &mut DbEntry,
+	github_bot: &GithubBot,
+	issue: &github::Issue,
+	admin_of: &(github::Project, project_info::ProjectInfo),
+) -> Result<DbEntryState> {
+	Ok(
+		if let Some(backlog_column) = admin_of
+			.0
+			.columns_url
+			.as_ref()
+			.and_then(|url| github_bot.project_columns(url).ok())
+			.and_then(|columns| {
+				columns.into_iter().find(|c| {
+					c.name
+						.as_ref()
+						.map(|name| {
+							name.to_lowercase()
+								== PROJECT_BACKLOG_COLUMN_NAME.to_lowercase()
+						})
+						.unwrap_or(false)
+				})
+			}) {
+			db_entry.issue_project_state =
+				Some(ProjectState::Confirmed(backlog_column.id));
+			github_bot.create_project_card(
+				backlog_column.id,
+				issue.id.context(error::MissingData)?,
+				github::ProjectCardContentType::Issue,
+			)?;
+			DbEntryState::Update
+		} else {
+			// TODO warn that the project lacks a backlog column
+			DbEntryState::DoNothing
+		},
+	)
+}
+
+fn author_core_no_project(
+	db_entry: &mut DbEntry,
 	github_bot: &GithubBot,
 	matrix_bot: &MatrixBot,
-	core_devs: &[github::User],
-	github_to_matrix: &HashMap<String, String>,
-	projects: Option<&project::Projects>,
 	issue: &github::Issue,
-) -> Result<()> {
-	// TODO: handle multiple projcets in a single repo
-	let project_info =
-		projects.and_then(|p| p.0.iter().last().map(|p| p.1.clone()));
+	default_channel_id: &str,
+	since: Option<Duration>,
+) -> Result<DbEntryState> {
+	let issue_id = issue.id.context(error::MissingData)?;
+	let issue_html_url = issue.html_url.as_ref().context(error::MissingData)?;
 
-	let author = &issue.user;
-	let repo = &issue.repository;
-	let author_info = project_info
-		.map_or_else(project::AuthorInfo::default, |p| {
-			p.author_info(&author.login)
-		});
-	let author_is_core = core_devs.iter().any(|u| u.id == author.id);
+	let ticks = since
+		.map(|elapsed| elapsed.as_secs() / ISSUE_NO_PROJECT_CORE_PING_PERIOD);
 
+	Ok(match ticks {
```

Comment:
I would recommend splitting this up so that you're not trying to handle every case in a single match statement. It makes it harder to read in my opinion. It be nice if the `None` was separate and done before this step so that you can write a more concise code that handles the actual `ticks`.

```rust
let ticks = match since.map(|elapsed| elapsed.as_secs() / ISSUE_NO_PROJECT_CORE_PING_PERIOD) {
    Some(ticks) => ticks,
    None => {
        db_entry.issue_no_project_ping = Some(SystemTime::now());
        matrix_bot.send_public_message(
            default_channel_id,
            &ISSUE_NO_PROJECT_MESSAGE.replace("{1}", issue_html_url),
        )?;
        return Ok(DbEntryState::Update);
    }
};

let state = if i == ISSUE_NO_PROJECT_ACTION_AFTER_NPINGS {
    // If after 3 days there is still no project
    // attached, move the issue to Core Sorting
    // repository
    github_bot.close_issue(&issue.repository.name, issue_id)?;
    let params = serde_json::json!({
            "title": issue.title,
            "body": issue.body.as_ref().unwrap_or(&"".to_owned())
    });
    github_bot.create_issue(CORE_SORTING_REPO, params)?;
    DbEntryState::Delete
} else if (db_entry.issue_no_project_npings) < i {
    db_entry.issue_no_project_npings = i;
    matrix_bot.send_public_message(
        default_channel_id,
        &ISSUE_NO_PROJECT_MESSAGE.replace("{1}", issue_html_url),
    )?;
    DbEntryState::Update
} else {
    DbEntryState::DoNothing
};

Ok(state)
```

Reply:
I'm getting rid of `DbEntryState` as it's too likely to break, so the point is moot, but in general I try to avoid early returns.

Point taken re: handling `ticks` specifically.